src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +526,437 @@ pub fn process_participation_record_updates<
+                // Since full base reward will be canceled out by inactivity penalty deltas,
+                // optimal participation receives full base reward compensation here.
+                rewards[i] += get_base_reward(state, i, context)?;
+            } else {
+                let reward_numerator =
+                    get_base_reward(state, i, context)? * (attesting_balance / increment);
+                rewards[i] += reward_numerator / (total_balance / increment);
+            }
+        } else {
+            penalties[i] += get_base_reward(state, i, context)?;
+        }
+    }
+    Ok((rewards, penalties))
+}
+
+pub fn get_source_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attester micro-rewards/penalties for source-vote for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let matching_source_attestations =
+        get_matching_source_attestations(state, previous_epoch, context)?;
+    get_attestation_component_deltas(state, matching_source_attestations, context)
+}
+
+pub fn get_target_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attester micro-rewards/penalties for target-vote for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let matching_target_attestations =
+        get_matching_target_attestations(state, previous_epoch, context)?
+            .cloned()
+            .collect::<Vec<_>>();
+    get_attestation_component_deltas(state, matching_target_attestations.as_slice(), context)
```

Comment:
so the types may get a little wonky, but ideally we avoid copying the `PendingAttestations`

they are owned by the `BeaconState` and (again ideally) we only pass in references to the right ones to `get_attestation_component_deltas`

i'd suggest doing the simple thing first to get it right, we can copy/clone/collect into `Vec` (or whatever is needed) to write the code and then can consider ways to avoid the copies where possible